src/mlpack/methods/decision_tree/gini_gain.hpp

```diff
@@ -44,37 +44,128 @@ class GiniGain
     if (labels.n_elem == 0)
       return 0.0;
 
-    // Count the number of elements in each class.
-    arma::vec counts(numClasses, arma::fill::zeros);
+    // Count the number of elements in each class.  Use four auxiliary vectors
```

Comment:
Yeah, this will be a primary place to look for optimizations: if SIMD instructions aren't being used where they could be, this is like a 4x slowdown in some cases.  Unfortunately digging deep to figure out what is being used is a little tedious.

I updated the code so that the information gain uses this approach also.